toolbox/src/base/WBBlock.cpp

```diff
@@ -0,0 +1,275 @@
+    Configuration config;
+    if (!getWBToolboxParameters(config, blockInfo)) {
+        return false;
+    }
+
+    // Check if the configuration is valid
+    if (!config.isValid()) {
+        Log::getSingleton().Error("The provided configuration is not valid.");
+        return false;
+    }
+
+    // Store the configuration inside the ToolboxSingleton
+    if (!interface.configure(confKey, config)) {
+        Log::getSingleton().Error("Failed to configure ToolboxSingleton.");
+        return false;
+    }
+
+    // Check if the DoFs are positive (and if the config has been stored successfully)
+    const Configuration& configFromSingleton = getConfiguration();
+
+    if (configFromSingleton.getNumberOfDoFs() < 1) {
+        Log::getSingleton().Error("Failed to configure WBBlock. Read 0 DoFs.");
+        return false;
+    }
+
+    return true;
+}
+
+bool WBBlock::initialize(BlockInformation* blockInfo)
+{
+    // CONFIGURE the ToolboxSingleton
+    // ==============================
+
+    ToolboxSingleton& interface = ToolboxSingleton::sharedInstance();
+
+    // Gather again confKey
+    // Note: despite data for blocks is not persistent between configureSizeAndPorts()
+    //       and initialize() (e.g. confKey), the singleton is not destroyed.
+    //       This means that the stored configurations are still there.
+    if (!blockInfo->getStringParameterAtIndex(PARAM_IDX_CONFBLOCKNAME, confKey)) {
+        Log::getSingleton().Error("Failed to retrieve the struct with parameters.");
+        return false;
+    }
+
+    // Check if the key is valid
+    if (!interface.isKeyValid(confKey)) {
+        Log::getSingleton().Error("Failed to retrieve the configuration object from the singleton.");
+        return false;
+    }
+
+    // Initialize the iDynTreeRobotState struct
+    unsigned dofs = interface.numberOfDoFs(confKey);
+    if (!initializeRobotState(dofs)) {
+        Log::getSingleton().Error("Failed to initialize the iDynTree's robot state.");
+        return false;
+    }
+
+    return true;
+}
+
+bool WBBlock::terminate(BlockInformation* /*blockInfo*/)
+{
+    // Deallocate iDynTreeRobotState memory
+    if (robotState.m_gravity) {
+        delete robotState.m_gravity;
+        robotState.m_gravity = nullptr;
+    }
+    if (robotState.m_jointsPosition) {
+        delete robotState.m_jointsPosition;
+        robotState.m_jointsPosition = nullptr;
+    }
+    if (robotState.m_world_T_base) {
+        delete robotState.m_world_T_base;
+        robotState.m_world_T_base = nullptr;
+    }
+    if (robotState.m_baseVelocity) {
+        delete robotState.m_baseVelocity;
+        robotState.m_baseVelocity = nullptr;
+    }
+    if (robotState.m_jointsVelocity) {
+        delete robotState.m_jointsVelocity;
+        robotState.m_jointsVelocity = nullptr;
+    }
+
+    // If the RemoteControlBoardRemapper has been allocated, free its memory
+    ToolboxSingleton& interface = ToolboxSingleton::sharedInstance();
```

Comment:
I think that this part should be managed by the specific subclasses.
If we do not explicitly `retain` in the initialise we must not call `release` in the terminate

Reply:
I agree, it is cleaner and less error prone. This must go in the new developers documentation.